ev3dev2/motor.py

```diff
@@ -1921,6 +1954,98 @@ def on(self, left_speed, right_speed):
+            is on the edge of the line.  If this is None we assume that the color sensor
+            is on the edge of the line and will take a reading to set this variable.
+
+        ``follow_left_edge`` determines if we follow the left or right edge of the line
+
+        ``WHITE`` is the reflected_light_intensity that is used to determine if we have
+            lost the line
+
+        ``OFF_LINE_COUNT`` is how many consecutive times through the loop the
+            reflected_light_intensity must be greater than ``WHITE`` before we
+            declare the line lost and raise an exception
+
+        ``SLEEP_SEC`` is how many seconds we sleep on each pass through
+            the loop.  This is to give the robot a chance to react
+            to the new motor settings. This should be something small such
+            as 0.01 (10ms).
+        """
+        assert self.cs, "ColorSensor must be defined"
+
+        if target_light_intensity is None:
+            target_light_intensity = self.cs.reflected_light_intensity
+
+        integral = 0.0
+        last_error = 0.0
+        derivative = 0.0
+        off_line_count = 0
+        speed_native_units = speed.to_native_units(self.left_motor)
+        MAX_SPEED = SpeedNativeUnits(self.max_speed)
+
+        while keep_following_function(self):
+            reflected_light_intensity = self.cs.reflected_light_intensity
+            error = target_light_intensity - reflected_light_intensity
+            integral = integral + error
+            derivative = error - last_error
+            last_error = error
+            turn_native_units = (kp * error) + (ki * integral) + (kd * derivative)
+
+            if follow_left_edge:
+                left_speed = SpeedNativeUnits(speed_native_units - turn_native_units)
+                right_speed = SpeedNativeUnits(speed_native_units + turn_native_units)
+
+            else:
+                left_speed = SpeedNativeUnits(speed_native_units + turn_native_units)
+                right_speed = SpeedNativeUnits(speed_native_units - turn_native_units)
+
+            if left_speed > MAX_SPEED:
+                log.info("%s: left_speed %s is greater than MAX_SPEED %s"  % (self, left_speed, MAX_SPEED))
+                self.stop()
+                raise LineFollowTooFast("The robot is moving too fast to follow the line")
+
+            if right_speed > MAX_SPEED:
+                log.info("%s: right_speed %s is greater than MAX_SPEED %s"  % (self, right_speed, MAX_SPEED))
+                self.stop()
+                raise LineFollowTooFast("The robot is moving too fast to follow the line")
+
+            # Have we lost the line?
+            if reflected_light_intensity >= WHITE:
+                off_line_count += 1
+
+                if off_line_count >= OFF_LINE_COUNT_MAX:
+                    self.stop()
+                    raise LineFollowFailed("we lost the line")
+            else:
+                off_line_count = 0
+
+            if SLEEP_SEC:
```

Comment:
No sleep at all by default? I would suggest making sure that there's always a consistent delay otherwise the derivative and integral behavior will by inconsistent and unpredictable. Alternately, we could use real units for derivative and integral (i.e., divide by time delta).

Reply:
I'll do some testing with 10ms and see how that behaves.  This is another one of those things that might be highly dependent on the type of robot.

Reply:
10ms seemed fine